### DISCUSSION TITLE
How does WinGet identify packages for applications installed on a system

### DISCUSSION THREAD

**Discussion by Daniel-Tr** [microsoft/winget-cli#3033] [Q&A] [answered]
I'm not sure if this is a bug or not, so I'll start with a 'simple' question: 

How exactly does the WinGet client identify packages for applications on the system, which have not been installed via WinGet?

Background for this: We have an application installed (TreeSize Professional), simply by runnign the installer. When running 'WinGet list' or 'WinGet upgrade --all', it shows an upgrade to be available for the JAMSoftware.TreeSize.Trial package. So there obviously has to be some sort of mapping in between the package and the installed application. Problem is that they are different (at least for now), so this mapping turns out wrong.

**Comment by Trenly** [comment, marked as answer]
Winget uses the information in the AppsAndFeaturesEntries to try and match to the metadata in a manifest. This may sound complicated, but really it's just using the information that's in the "Add & Remove Programs" section of control panel. I believe in this case, it is trying to match only on "TreeSize" instead of the full display name, which can likely be fixed by updating the metadata for the Manifest at winget-pkgs

**Reply by Daniel-Tr** [reply]
So it's basically just guessing it, depending on matching substrings? Doesn't that create security concerns, when using this method? And how would a user have to deal with false matches? Or the maintainer of the package in WinGet?

I tried to update the package, explicitly mismatching the installer information with the data written to the registry/"Add & Remove Programs". As I said: WinGet is installing a trial version, erroneously matching and overwriting the full version. 
You dismissed the pull request, stating the information "MUST match". But if so, I feel like I have to demand, that the package MUST match the installer as well - which it simply doesn't. And the user has no option yet to avoid it (even though the PIN command seems to be in development).

I will issue a new PR, temporarily deleting the package. We don't want our customers to face this problem, so I hope you can accept this request. We will reintroduce the package with the upcoming V9, incorporating a fix on our side.

**Reply by denelon** [reply]
These challenges are largely due to how "installers" report themselves via Windows Apps & Features. If a "stable" version of a package doesn't have a different "displayName" from a "preview" version, then WinGet has no mechanism to determine the difference.

We've been categorizing several issues with the "Area-Matching" label to start addressing some of these issues, but there are still some other subtle challenges with packages that install "Side-by-Side" rather than actually updating the "current" package.

**Reply by Daniel-Tr** [reply]
Ok, thank you very much for the additional info!

**Reply by e-t-l** [reply]
> due to how "installers" report themselves via Windows Apps & Features

How *do* installers report themselves? Is this stored in the Registry or queried via some system API?

**Reply by denelon** [reply]
It's stored in registry keys. Various installer tools handle most of the heavy lifting. MSIX is a bit different from MSI-based and .exe based installers, so it's not really a "one size fits all" kind of thing. When you look in Windows Settings at "Apps" the metadata displayed there is coming from the registry as well as the "uninstall key" that get used by WinGet for List, Uninstall, and Upgrade.

**Reply by e-t-l** [reply]
Ah ok thank you, I think that answered what I was really interested in: Winget uses the uninstall registry keys to determine what programs are currently installed.

**Reply by bryandam** [reply]
@denelon, sorry to necro-post, but this is a top search result and I'd like to ask a clarifying question.

How does this work for packages that do not have the AppsAndFeaturesEntries metadata?
For example, lucasdeeiroz.RobotRunner 2.2.29: [https://github.com/microsoft/winget-pkgs/tree/master/manifests/l/lucasdeeiroz/RobotRunner/2.2.29](https://github.com/microsoft/winget-pkgs/tree/master/manifests/l/lucasdeeiroz/RobotRunner/2.2.29)

Presumably it's some similar logic based on Package Name, Publisher, and Version? If so, since Package Name comes from the locale, presumably it would match on the default locale as well as any locale specific Package Names?

**Reply by denelon** [reply]
The installer does appear to write data to the registry when it gets installed:

<img width="1718" height="736" alt="image" src="https://github.com/user-attachments/assets/a815e5b7-b92b-4c2d-b01c-652fd70f9a88" />

**Reply by bryandam** [reply]
Right, what I'm trying to understand a bit more in-depth is the mechanics of what metadata is used to do that and what the logic is since that packages does not have an explicit AppsAndFeaturesEntries element that @Trenly mentions.

It _feels_ as if it's looking at the default locale manifest and using the Package Name, Version, and Publisher and trying to match all three to the User/Machine ARP entries? Are they fuzzy searches (ex. implied wildcards at beginning or end) or must it be an exact match?

**Reply by denelon** [reply]
Yes, we're attempting to correlate using the publisher name and the package  name.

There are some heuristics used, and there are values routinely used at the end of the "displayName" in parenthesis that often get ignored.
